contracts/distribution/dao-rewards-distributor/src/state.rs

```diff
@@ -54,41 +78,119 @@ pub struct DenomRewardState {
     pub funded_amount: Uint128,
     /// optional destination address for reward clawbacks
     pub withdraw_destination: Addr,
+    /// historic denom distribution epochs
+    pub historic_epoch_configs: Vec<EpochConfig>,
 }
 
 impl DenomRewardState {
-    pub fn bump_last_update(mut self, current_block: &BlockInfo) -> Self {
-        self.last_update = match self.emission_rate.duration {
+    /// Sum all historical total_earned_puvp values.
+    pub fn get_historic_rewards_earned_puvp_sum(&self) -> Uint256 {
+        self.historic_epoch_configs
+            .iter()
+            .fold(Uint256::zero(), |acc, epoch| acc + epoch.total_earned_puvp)
+    }
+
+    /// Finish current epoch early and start a new one with a new emission rate.
+    pub fn transition_epoch(
+        &mut self,
+        new_emission_rate: RewardEmissionRate,
+        current_block: &BlockInfo,
+    ) -> StdResult<()> {
+        let current_block_expiration = match self.active_epoch_config.emission_rate.duration {
+            Duration::Height(_) => Expiration::AtHeight(current_block.height),
+            Duration::Time(_) => Expiration::AtTime(current_block.time),
+        };
+
+        // 1. finish current epoch by changing the end to now
+        let mut curr_epoch = self.active_epoch_config.clone();
+        curr_epoch.ends_at = current_block_expiration;
+        curr_epoch.finish_block = Some(current_block.to_owned());
+
+        // TODO: remove println
+        println!("transition_epoch: {:?}", curr_epoch);
+        // 2. push current epoch to historic configs
+        self.historic_epoch_configs.push(curr_epoch.clone());
+
+        // 3. deduct the distributed rewards amount from total funded amount,
+        // as those rewards are no longer available for distribution
+        let curr_epoch_earned_rewards = match curr_epoch.emission_rate.amount.is_zero() {
+            true => Uint128::zero(),
+            false => self.active_epoch_config.get_total_rewards()?,
+        };
+        self.funded_amount = self.funded_amount.checked_sub(curr_epoch_earned_rewards)?;
```

Comment:
```suggestion
        // 1. finish current epoch by changing the end to now
        let mut curr_epoch = self.active_epoch_config.clone();
        curr_epoch.ends_at = current_block_expiration;
        curr_epoch.finish_block = Some(current_block.to_owned());

        // TODO: remove println
        println!("transition_epoch: {:?}", curr_epoch);
        // 2. push current epoch to historic configs
        self.historic_epoch_configs.push(curr_epoch.clone());

        // 3. deduct the distributed rewards amount from total funded amount,
        // as those rewards are no longer available for distribution
        let curr_epoch_earned_rewards = match curr_epoch.emission_rate.amount.is_zero() {
            true => Uint128::zero(),
            false => curr_epoch.get_total_rewards()?,
        };
        self.funded_amount = self.funded_amount.checked_sub(curr_epoch_earned_rewards)?;
```

i'm pretty sure this should be using `curr_epoch`, right? since you're trying to remove the rewards already distributed here, and `curr_epoch` contains the updated `ends_at` from step 1.

Reply:
that makes sense to me!

Reply:
sweet!